src/main/java/com/rcv/Tabulator.java

```diff
@@ -775,45 +741,54 @@ private boolean hasContinuingCandidates(SortedMap<Integer, Set<String>> rankToCa
-            cvr.addRoundOutcome(
-                VoteOutcomeType.COUNTED, selectedCandidateID, fractionalTransferValue);
-
-            // Increment the tally for this candidate by the fractional transfer value of the CVR.
-            incrementTally(roundTally, fractionalTransferValue, selectedCandidateID);
-            // We set this in case we need to redistribute votes if this is a multi-winner race and
-            // this candidate wins, but there are still more winners to come.
-            cvr.setCurrentRecipientOfVote(selectedCandidateID);
-
-            if (config.isTabulateByPrecinctEnabled()
-                && cvr.getPrecinct() != null
-                && !cvr.getPrecinct().isEmpty()) {
-              incrementTally(
-                  roundTallyByPrecinct.get(cvr.getPrecinct()),
-                  fractionalTransferValue,
-                  selectedCandidateID);
-            }
+        for (String candidate : candidateSet) {
+          if (!isCandidateContinuing(candidate)) {
+            continue;
+          }
+          // If this fails, it means the code failed to handle an overvote with multiple
+          // continuing candidates.
+          assert selectedCandidate == null;
+          // we found a continuing candidate, so increase their tally by 1
+          selectedCandidate = candidate;
+          // the FTV for this cast vote record (by default the FTV is exactly one vote, but it
+          // could be less in a multi-winner contest if this CVR already helped elect a winner.)
+          BigDecimal fractionalTransferValue = cvr.getFractionalTransferValue();
+          cvr.addRoundOutcome(VoteOutcomeType.COUNTED, selectedCandidate, fractionalTransferValue);
+
+          // Increment the tally for this candidate by the fractional transfer value of the CVR.
+          incrementTally(roundTally, fractionalTransferValue, selectedCandidate);
+          // We set this in case we need to redistribute votes if this is a multi-winner race and
+          // this candidate wins, but there are still more winners to come.
+          cvr.setCurrentRecipientOfVote(selectedCandidate);
+
+          if (config.isTabulateByPrecinctEnabled()
+              && cvr.getPrecinct() != null
+              && !cvr.getPrecinct().isEmpty()) {
+            incrementTally(
+                roundTallyByPrecinct.get(cvr.getPrecinct()),
+                fractionalTransferValue,
+                selectedCandidate);
           }
         }
 
-        if (selectedCandidateID != null && !selectedCandidateID.isEmpty()) {
+        if (selectedCandidate != null) {
           // we've found our candidate
           break;
         }
       } // end looping over the rankings within one ballot
+
+      if (!cvr.isExhausted() && cvr.getCurrentRecipientOfVote() == null) {
+        // Either it's an undervote or their candidates got eliminated.
+        if (config.getMaxSkippedRanksAllowed() != null
```

Comment:
Unfortunately I don't follow this logic 100%, but I'll trust that it's good. More reason to get our automated tests going ASAP (hello again, #96).

Reply:
Yeah, it's pretty complicated. I'll add this comment inline:

Once we get here (having iterated through some or all of this CVR's rankings), either:
a) the ballot is exhausted,
b) we've counted it toward a candidate, or
c) neither.
If (c) is the case, we definitely want to exhaust the ballot, and the only question is whether the official reason for exhaustion: did it skip too many rankings at the end of the ballot, or did it just no longer have any continuing candidates?